master/internal/templates/api.go

```diff
@@ -0,0 +1,259 @@
+}
+
+// GetTemplate by name. Returns an error if the requested template does not exist.
+func (a *TemplateAPIServer) GetTemplate(
+	ctx context.Context,
+	req *apiv1.GetTemplateRequest,
+) (*apiv1.GetTemplateResponse, error) {
+	user, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Unauthenticated, "failed to get the user: %s", err)
+	}
+	if len(req.TemplateName) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "name is required")
+	}
+
+	var t templatev1.Template
+	err = db.Bun().NewRaw(`
+SELECT name, config, workspace_id
+FROM templates
+WHERE name = ?
+`, req.TemplateName).Scan(ctx, &t)
+	switch {
+	case errors.Is(err, sql.ErrNoRows):
+		return nil, api.NotFoundErrs("template", req.TemplateName, true)
+	case err != nil:
+		return nil, fmt.Errorf("fetching template %s from database: %w", req.TemplateName, err)
+	}
+
+	permErr, err := AuthZProvider.Get().CanViewTemplate(ctx, user, model.AccessScopeID(t.WorkspaceId))
+	switch {
+	case err != nil:
+		return nil, status.Errorf(codes.Internal, "failed to check for permissions: %s", err)
+	case permErr != nil:
+		return nil, authz.SubIfUnauthorized(permErr, api.NotFoundErrs("template", req.TemplateName, true))
+	}
+	return &apiv1.GetTemplateResponse{Template: &t}, nil
+}
+
+// PutTemplate creates or updates a template.
+func (a *TemplateAPIServer) PutTemplate(
+	ctx context.Context,
+	req *apiv1.PutTemplateRequest,
+) (*apiv1.PutTemplateResponse, error) {
+	if len(req.Template.Name) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "name is required")
+	}
+	if req.Template.WorkspaceId != 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "setting workspace_id is not supported.")
+	}
+
+	if _, err := TemplateByName(ctx, req.Template.Name); err != nil {
+		_, err := a.PostTemplate(ctx, &apiv1.PostTemplateRequest{Template: req.Template})
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		_, err = a.PatchTemplateConfig(
+			ctx,
+			&apiv1.PatchTemplateConfigRequest{
+				TemplateName: req.Template.Name,
+				Config:       req.Template.Config,
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &apiv1.PutTemplateResponse{Template: req.Template}, nil
+}
+
+// PostTemplate creates a template. If a template with the same name exists, an error is returned.
+func (a *TemplateAPIServer) PostTemplate(
+	ctx context.Context,
+	req *apiv1.PostTemplateRequest,
+) (*apiv1.PostTemplateResponse, error) {
+	user, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Unauthenticated, "failed to get the user: %s", err)
+	}
+	if len(req.Template.Name) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "name is required")
+	}
+
+	workspaceID := int(req.Template.WorkspaceId)
+	if req.Template.WorkspaceId == 0 {
+		workspaceID = model.DefaultWorkspaceID
+	}
+
+	// TODO(DET-9839): This random workspaces query should be in a workspaces package.
+	exists, err := db.Bun().NewSelect().Table("workspaces").
```

Comment:
this query leaks what workspaces exist and don't in RBAC


We need a can view workspace check here?